
Nama : M. Hafiz Attariq<br>
NIM : 24060121140174<br>

# Import Libraries

In [2]:
# Import Libraries
import pandas as pd
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
import seaborn as sns

# Memuat Dataset

In [3]:
# Memuat Dataset
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']

dataset = pd.read_csv(url, names=names)

# Dimensi Dataset

In [ ]:
dataset.shape

(150, 5)

jumlah baris dalam dataset adalah 12960 dan jumlah kolom dalam data set adalah 9

# Isi Dataset

In [ ]:
dataset

sepal-length  sepal-width  petal-length  petal-width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

Diatas merupakan tabel dari isi dataset

# Distribusi Kelas Data


In [ ]:
dataset.groupby('class').size()

class
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
dtype: int64

### Pertemuan 2

## Import Model Selection

In [4]:
from sklearn import model_selection

##Train-Test Split

In [5]:
array = dataset.values

X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

X berisi fitur (features) dari dataset, yaitu kolom 1 hingga 4.<br>
Y berisi target (label) dari dataset, yaitu kolom 5.<br><br>
train_test_split adalah fungsi dari modul model_selection yang digunakan untuk membagi dataset menjadi data latih dan data uji.<br>
X_train dan Y_train adalah data latih untuk fitur dan target.<br>
X_validation dan Y_validation adalah data uji untuk fitur dan target.<br>
test_size=0.20 menunjukkan bahwa 20% dari dataset akan digunakan sebagai data uji.<br>
random_state digunakan untuk mengatur seed pada proses pembagian data agar hasilnya dapat direproduksi.

##  Menampilkan shape dari masing-masing subset data setelah dilakukan pembagian.

In [6]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


## K-Folds Cross Validation

In [7]:
seed = 7

scoring = 'accuracy'

## Model Algoritma di Modul

In [ ]:
# #Import library algoritma
# from sklearn.neighbors import KNeighborsClassifier
# from sklearn.svm import SVC

# #Spot Check Algorithms
# models = []

# models.append(('KNN', KNeighborsClassifier()))
# models.append(('NB', GaussianNB()))
# models.append(('SVM', SVC()))

## Eksplorasi Model Algoritma Lain

In [33]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier

models = []

models.append(('Random Forest', RandomForestClassifier()))
models.append(('Gradient Boosting', GradientBoostingClassifier()))
models.append(('Decision Tree', DecisionTreeClassifier()))

## Evaluasi Algoritma Model

In [34]:
# evaluate each model in turn
results = []

names = []
for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train,
  cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

Random Forest: 0.966667 (0.040825)
Gradient Boosting: 0.950000 (0.055277)
Decision Tree: 0.958333 (0.076830)


Mengevaluasi kinerja beberapa model machine learning menggunakan validasi silang (cross-validation).<br>
Model yang dinilai:<br>


*   Random Forest: Akurasi rata-rata sekitar 96.67%, dengan deviasi standar sekitar 0.040825.
*   Gradient Boosting: Akurasi rata-rata sekitar 95.00%, dengan deviasi standar sekitar 0.055277.
*   Decision Tree: Akurasi rata-rata sekitar 95.83%, dengan deviasi standar sekitar 0.076830.





## Import Library untuk Testing

In [35]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

## Melakukan pengujian tentang keakuratan model Random Forest terhadap data

In [37]:
# Make predictions on validation dataset
RandomForest = RandomForestClassifier()

RandomForest.fit(X_train, Y_train)

predictions = RandomForest.predict(X_validation)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8666666666666667
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



karena Random Forest memiliki nilai akurasi perkiraan terbesar, maka dilakukan pengujian mengenai keakuratan Model tersebut.<br>
Akurasi model pada dataset validasi adalah sekitar 86.67%.